Allow dumpargs to format bit flags

LorenVS opened the pull request:
If flags are supplied for a function in `functions_data.py`, the `dumpargs` command can pretty-print the combination of flags used in the call.

This is some preliminary work on #2734. I wanted to get some initial feedback before looking at how to populate the correct flag values for all the different calls that are currently present in `functions_data.py`.

Commit:
Allow dumpargs to format bit flags

If flags are supplied for a funciton in `functions_data.py`, the
`dumpargs` command can pretty-print the combination of flags used in the
call.

disconnect3d commented on tests/gdb-tests/tests/test_commands_dumpargs.py:
```diff
@@ -0,0 +1,40 @@
+    assert out[0] == '        addr:      0xcafe0000'
+    assert out[1] == '        len:       0x1000'
+    assert out[2] == '        prot:      1'
+    assert out[3] == '        flags:     MAP_PRIVATE | MAP_ANONYMOUS | MAP_FIXED'
```
Maybe: `flags:     0x32 (MAP_PRIVATE|MAP_ANONYMOUS|MAP_FIXED)`?

LorenVS replied:
Agreed, it makes complete sense to retain the original integer value as well. I updated the format to this.

Commit:
Update format for flags arguments.

The format now includes the original integer value *and* the extracted
flag names (if any). A sample output would be `0x03 (FLAG_2|0x01)`.

disconnect3d committed:
Fix lint

disconnect3d commented on pwndbg/commands/dumpargs.py:
```diff
@@ -38,7 +38,7 @@ def call_args() -> List[str]:
     for arg, value in pwndbg.arguments.get(pwndbg.aglib.disasm.one()):
         code = arg.type != "char"
-        pretty = pwndbg.chain.format(value, code=code)
+        pretty = pwndbg.chain.format(value, code=code, flags=arg.flags)
```
Hmm... does it make sense to call `chain.format` at all if flags are there? If I understand correctly, when we pass a list of values, we will never pass flags; when we pass flags, we will never want to do a "chain dereference".

Maybe this should be sth like:
```py
pretty = pwndbg.chain.format(value, code=code) if not args.flags else format_flags(value, flags)
```

(The `format_flags` should probably be put in some module instead of here)

LorenVS replied:
I had a hard time determining whether this logic belonged in `chain.format` or elsewhere. I switched to your suggestion. I didn't see a great place to put `format_flags_argument`, so I moved it into `functions.py`.

disconnect3d commented on pwndbg/chain.py:
```diff
@@ -181,3 +186,37 @@ def format(
+    #
+    # For example:
+    # 0x03 (FLAG_2|0x01)
+    return hex(original_value) + " (" + '|'.join(flag_names) + ")"
```
This should be:
```suggestion
    return f"{original_value:#x} ({'|'.join(flag_names)})"
```

disconnect3d commented on pwndbg/chain.py:
```diff
@@ -181,3 +186,37 @@ def format(
 
     return arrow_right.join(rest) + arrow_left + enhanced
+
+def _format_flags_value(flags: List[Flag], value: int):
```
Ideally, we should add some tests for this function

LorenVS replied:
Done

disconnect3d commented on pwndbg/lib/functions_data.py:
```diff
@@ -28479,7 +28480,13 @@ def load_functions():
             Argument(type="size_t", derefcnt=0, name="len"),
             Argument(type="int", derefcnt=0, name="prot"),
-            Argument(type="int", derefcnt=0, name="flags"),
+            Argument(type="int", derefcnt=0, name="flags", flags=[
```
Should always be a tuple instead of a list

LorenVS replied:
I switched the type everywhere to `tuple[Flag, ...]`. I think that is what you're asking for?

disconnect3d commented:
Pls fix linter issues: https://github.com/pwndbg/pwndbg/actions/runs/13348834380/job/37282697043?pr=2740 and see comments :)

I like this feature a lot!

Commit:
Address review comments

Commit:
Remove import after rebase

LorenVS commented:
Sorry for the lint failures, I had a difficult time getting them up and running locally at first. I think I have it figured out now.

disconnect3d reviewed (approved):
Awesome, thanks!

OBarronCS commented:
This is awesome!

Do we happen to have knowledge as to how the `functions_data.py` file was initially populated? It looks like an auto-generated file. Some of the argument names for the functions don't match exactly with what manpages show, for example.

OBarronCS commented:
Just thought of something to consider in the future as well:

Looks like these flag constants actually differ based on the target os/arch pair:
(searching pwntools codebase for `#define MAP_ANONYMOUS` for example - it pulls these constants from linux sourcecode)
![image](https://github.com/user-attachments/assets/7dbb6ac9-5bfb-41b7-9676-496b0af753c4)

We could potentially dynamically pull them - look into the pwntools `constgrep` tool for some details.

patryk4815 commented:
> We could potentially dynamically pull them - look into the pwntools `constgrep` tool for some details.

eg using:
```py
# ./result/bin/pwndbg
pwndbg> ipi
In [2]: from pwnlib import constants
In [3]: constants.MAP_ANONYMOUS
Out[3]: Constant('MAP_ANONYMOUS', 0x20)

# ./result/bin/pwndbg --init-eval-command='set architecture mips'
The target architecture is set to "mips".
pwndbg> ipi

In [1]: from pwnlib import constants
In [2]: constants.MAP_ANONYMOUS
Out[2]: Constant('MAP_ANONYMOUS', 0x800)
```

LorenVS commented:
It looks like two commits from this PR (3fa6409c4a6590e8fc25bfd744aa2ec009fd9565 and 0a235d0f0015266a5188674c2be9cb8926900661) were not merged. The merged version contains the pre-review-comments implementation.

LorenVS commented:
Thanks for making me aware of `pwnlib`, that makes the follow-up work here trivial. I sent out PR #2755 to address the os/arch issues.
